web/src/reducers/budget.js

```diff
@@ -250,113 +254,172 @@ const buildBudget = bigState => {
       });
     });
 
-    /**
-     * Get the amount of "other funding" attributed to each type of
-     * cost for a given year of the current activity.
-     * @description For each type of cost, this method calculates that type's
-     *    contribution to the total cost for the year.  It then applies that
-     *    same percentage to the total "other funding".  For example:
-     *
-     *    other funding is $200
-     *    the total cost is $1,000
-     *    contractor costs are $500
-     *    contractor costs is 50% of total costs
-     *    so other funding for contractor costs is 50% * $200 = $100
-     *
-     *    This function is placed here in the code because it depends on the
-     *    results of the previous calculations.  This way if anyone accidentally
-     *    uses it earlier in the code, it will be an error because the function
-     *    hasn't been defined yet.
-     * @param {String} year As a four-character year string (e.g., '2018')
-     * @returns {{contractors:Number, expenses:Number, statePersonnel:Number}} Other funding for each type of cost
-     */
-    const getOtherFunding = year => {
-      const otherTotal = n(allocation[year].other);
-
-      const costs = newState[fundingSource];
-      const yearTotal = costs.combined[year].total;
-
-      return {
-        contractors: otherTotal * costs.contractors[year].total / yearTotal,
-        expenses: otherTotal * costs.expenses[year].total / yearTotal,
-        statePersonnel:
-          otherTotal * costs.statePersonnel[year].total / yearTotal
-      };
-    };
+    // Converts an array where the values correspond to contractors,
+    // expenses, and state personnel values, respectively.
+    const costShareReducer = (acc, value, i) => ({
+      ...acc,
+      [['contractors', 'expenses', 'statePersonnel'][i]]: value
+    });
 
     // Now loop back over the years and compute state and federal shares
     // of all the costs.
     years.forEach(year => {
-      const other = getOtherFunding(year);
+      const totalCost = activityTotals.data.combined[year];
+      const totalGovShare = totalCost - n(allocation[year].other);
+      const costShares = getAllocation(year, totalGovShare);
+
+      // This represents the percentage each cost category contributes to the
+      // total activity cost.  This is useful for distributing the total
+      // in a way that we get whole numbers and preserve sums.
+      const costCategoryPercentages = [
+        activityTotals.data.contractors[year] / totalCost,
+        activityTotals.data.expenses[year] / totalCost,
+        activityTotals.data.statePersonnel[year] / totalCost
+      ];
+
+      // Compute the actual cost of each cost category, based on the federal
+      // and state shares and the percentages calculated above.
+      const fedShare = roundedPercents(
+        costShares.fedShare,
```

Comment:
The comment on this piece of code isn't great (probably made sense when I was in the weeds, but now, not so much), but yes, it's okay here.  The `costShares` is being divvied up into cost categories, and those cost categories will sum back to `costShares`.  It's...  confusing, because we're calculating a percent of a percent.  But I think a tree helps me understand it:

```
100%                                _____ $1,500 ____
                                  /                  \
                                 /                    \
90%/10%                   $1,350                        $150
federal/state            /   |  \                     /   |  \
                        /    |   \                   /    |   \
23%/42%/35%        $310.50  $567  $472.50        $34.50  $63  $52.50
personnell/        $311     $567  $472           $35     $63  $52
contractor/
other
```

(that took longer than I want to admit...)

Anyway, what's happening is that `costShares` isn't being rounded again - it's being used as the basis for rounding the cost of each cost category, such that the cost categories sum up to the `costShares`, which in turn sum up to the total Medicaid cost.

It's a little convoluted, but it's because of the "other funding."  It comes off the top, not out of any particular cost category, but CMS wants to see where their dollars are going.  In order to be able to show that, we have to deduct the "other funding" amount proportionally from each cost category.  😬  We actually used to do it that way - we figured the proportions, divided up the "other funding" accordingly, and subtracted it from each cost category total.  But it turns out that once you have the proportions, it's a lot simpler to take the other funding off the total and then divvy up the Medicaid total proportionally instead.

I don't know if this is making anything clearer.  The best I can offer is that we can spend a half hour some time walking through all the numbers in the app to show where they came from, if that'd be helpful.  There are a bunch of them!